File: src/bgbg/engine/objects.py
"""The segmentation overlay model: objects, selection, hover focus. Stdlib only.

What the canvas shows on top of the image — the per-object masks, their colours
and contours, which are selected, which is focused, and the outline tween between
two focused objects.

The engine never holds a texture. Each object carries an opaque `handle` that the
shell minted when it decoded the mask (a Gdk.Texture on the desktop, an
ImageBitmap on the web); the engine only ever passes it back in a display list.
The handle is the object's id, which is what makes the display-list goldens
readable.

The parallel dicts (masks/colors/polys/...) are deliberate rather than a list of
object records: `scene()` runs every frame and this way it only rebinds
references, instead of rebuilding up to 256 objects at 60fps.
"""
from .anim import MORPH_N
from .color import parse_color
from .geometry import polygon_area_abs, resample_closed, align_ring
from .hittest import HitMaps
from .render.scene import Scene, Morph, POINT_COLOR


class ObjectStore:
    def __init__(self):
        self.seg_mode = None          # None | "everything" | "point"

        self.masks = {}               # oid -> handle   (insertion order = paint order)
        self.colors = {}              # oid -> (r,g,b,a)
        self.polys = {}               # oid -> [poly]   all contours
        self.sec_polys = {}           # oid -> [poly]   the non-largest contours
        self.centroids = {}           # oid -> (cx, cy) image px
        self.radius = {}              # oid -> ripple reach, image px
        self.rings = {}               # oid -> [N (x,y)] canonical outline ring

        self.selected = set()
        self.hover_id = 0             # the focused object (glow + outline)
        self.hover_gen = 0            # the general object under the cursor
        self.hover_spec = 0           # the specific one
        self.hover_depth = 0          # how many overlap here

        self.point_mask = None        # handle, point mode
        self.point_polys = ()

        self.morph = None             # Morph | None — an in-flight outline tween
        self.maps = HitMaps()

        # result-panel cutout preview: the source clipped to a union of masks
        self.clip_active = False
        self.clip_masks = ()
        self.clip_bg = None

    # ---------- loading ----------
    def load(self, objects):
        """Replace the objects. Each is `{id, color, contour, bbox, handle}` —
        the shell has already decoded the mask and minted `handle`, dropping any
        object whose mask failed to load.
        """
        self.clear_objects()
        for o in objects:
            oid = o["id"]
            self.masks[oid] = o.get("handle", oid)
            self.colors[oid] = parse_color(o["color"])
            contours = [[tuple(p) for p in poly] for poly in (o.get("contour") or [])]
            self.polys[oid] = contours
            bx, by, bw, bh = o.get("bbox", (0, 0, 0, 0))
            self.centroids[oid] = (bx + bw / 2.0, by + bh / 2.0)
            self.radius[oid] = 0.6 * max(bw, bh, 8)
            # The canonical ring (the largest polygon) is what the focus morph
            # lerps; the rest (holes, detached parts) crossfade during a switch.
            if contours:
                largest = max(contours, key=polygon_area_abs)
                self.rings[oid] = align_ring(resample_closed(largest, MORPH_N))
                self.sec_polys[oid] = [p for p in contours if p is not largest]

    def set_point_mask(self, handle, contour=None):
        self.point_mask = handle
        self.point_polys = [[tuple(p) for p in poly] for poly in (contour or [])]

    def set_composite(self, masks, bg=None):
        self.clip_active = True
        self.clip_masks = tuple(masks)
        self.clip_bg = bg

    def clear_composite(self):
        self.clip_active = False
        self.clip_masks = ()
        self.clip_bg = None

    def clear_objects(self):
        for d in (self.masks, self.colors, self.polys, self.sec_polys,
                  self.centroids, self.radius, self.rings):
            d.clear()

    def clear(self, keep_mode=False):
        self.clear_objects()
        self.selected.clear()
        self.hover_id = self.hover_gen = self.hover_spec = self.hover_depth = 0
        self.point_mask = None
        self.point_polys = ()
        self.morph = None
        self.maps.clear()
        if not keep_mode:
            self.seg_mode = None

    # ---------- selection ----------
    def selection(self):
        return sorted(self.selected)

    def set_selection(self, ids):
        self.selected = set(ids)

    def toggle(self, oid):
        """Returns True when the object was *added* (so the host can pop it)."""
        if oid in self.selected:
            self.selected.discard(oid)
            return False
        self.selected.add(oid)
        return True

    # ---------- state ----------
    def has_seg(self):
        return bool(self.masks) or self.point_mask is not None

    def active(self):
        """Overlay state that wants the frame clock running but that no clock
        owns — a hovered object, a selection, a point mask."""
        return bool(self.hover_id or self.selected
                    or self.point_mask is not None)

    def build_morph(self, old, new):
        """The tween between two focused objects, or None when there is nothing
        to tween (neither has an outline)."""
        fr = self.rings.get(old)
        to = self.rings.get(new)
        if fr is None and to is None:
            return None
        return Morph(
            from_ring=fr, to_ring=to,
            cf=self.colors.get(old) or self.colors.get(new) or POINT_COLOR,
            ct=self.colors.get(new) or self.colors.get(old) or POINT_COLOR,
            sec_from=self.sec_polys.get(old),
            sec_to=self.sec_polys.get(new),
            cen_from=self.centroids.get(old),
            cen_to=self.centroids.get(new))

    # ---------- the frame ----------
    def scene(self, image, image_size):
        sc = Scene()
        sc.image = image
        sc.image_size = image_size
        sc.seg_mode = self.seg_mode
        sc.masks = self.masks
        sc.colors = self.colors
        sc.polys = self.polys
        sc.sec_polys = self.sec_polys
        sc.centroids = self.centroids
        sc.radius = self.radius
        sc.selected = frozenset(self.selected)
        sc.hover_id = self.hover_id
        sc.hover_gen = self.hover_gen
        sc.hover_spec = self.hover_spec
        sc.hover_depth = self.hover_depth
        sc.point_mask = self.point_mask
        sc.point_polys = self.point_polys
        sc.morph = self.morph
        sc.clip_active = self.clip_active
        sc.clip_masks = self.clip_masks
        sc.clip_bg = self.clip_bg
        return sc

File: web/core/objects.ts
/** The segmentation overlay model — the mirror of engine/objects.py. */
import { MORPH_N } from "./anim";
import { parseColor, type RGBA } from "./color";
import {
  polygonAreaAbs,
  resampleClosed,
  alignRing,
  type Poly,
  type Vec2,
} from "./geometry";
import { HitMaps } from "./hittest";
import { Scene, Morph, POINT_COLOR } from "./render/scene";
import type { Handle } from "./render/ops";

export interface SegObjectSpec {
  id: number;
  color: string;
  contour?: number[][][];
  bbox?: [number, number, number, number];
  handle?: Handle;
}

export class ObjectStore {
  segMode: string | null = null;

  masks = new Map<number, Handle>();
  colors = new Map<number, RGBA>();
  polys = new Map<number, Poly[]>();
  secPolys = new Map<number, Poly[]>();
  centroids = new Map<number, Vec2>();
  radius = new Map<number, number>();
  rings = new Map<number, Poly>();

  selected = new Set<number>();
  hoverId = 0;
  hoverGen = 0;
  hoverSpec = 0;
  hoverDepth = 0;

  pointMask: Handle | null = null;
  pointPolys: Poly[] = [];

  morph: Morph | null = null;
  maps = new HitMaps();

  clipActive = false;
  clipMasks: Handle[] = [];
  clipBg: RGBA | null = null;

  load(objects: SegObjectSpec[]) {
    this.clearObjects();
    for (const o of objects) {
      const oid = o.id;
      this.masks.set(oid, o.handle ?? oid);
      this.colors.set(oid, parseColor(o.color)!);
      const contours: Poly[] = (o.contour ?? []).map(
        (poly) => poly.map((p) => [p[0], p[1]] as Vec2),
      );
      this.polys.set(oid, contours);
      const [bx, by, bw, bh] = o.bbox ?? [0, 0, 0, 0];
      this.centroids.set(oid, [bx + bw / 2.0, by + bh / 2.0]);
      this.radius.set(oid, 0.6 * Math.max(bw, bh, 8));
      if (contours.length) {
        // `max(..., key=area)` keeps the FIRST maximum on a tie in Python.
        let largest = contours[0];
        let best = polygonAreaAbs(largest);
        for (let i = 1; i < contours.length; i++) {
          const a = polygonAreaAbs(contours[i]);
          if (a > best) {
            best = a;
            largest = contours[i];
          }
        }
        this.rings.set(oid, alignRing(resampleClosed(largest, MORPH_N)));
        this.secPolys.set(
          oid,
          contours.filter((p) => p !== largest),
        );
      }
    }
  }

  setPointMask(handle: Handle | null, contour?: number[][][]) {
    this.pointMask = handle;
    this.pointPolys = (contour ?? []).map(
      (poly) => poly.map((p) => [p[0], p[1]] as Vec2),
    );
  }

  setComposite(masks: Handle[], bg: RGBA | null = null) {
    this.clipActive = true;
    this.clipMasks = [...masks];
    this.clipBg = bg;
  }

  clearComposite() {
    this.clipActive = false;
    this.clipMasks = [];
    this.clipBg = null;
  }

  clearObjects() {
    this.masks.clear();
    this.colors.clear();
    this.polys.clear();
    this.secPolys.clear();
    this.centroids.clear();
    this.radius.clear();
    this.rings.clear();
  }

  clear(keepMode = false) {
    this.clearObjects();
    this.selected.clear();
    this.hoverId = this.hoverGen = this.hoverSpec = this.hoverDepth = 0;
    this.pointMask = null;
    this.pointPolys = [];
    this.morph = null;
    this.maps.clear();
    if (!keepMode) this.segMode = null;
  }

  selection(): number[] {
    return [...this.selected].sort((a, b) => a - b);
  }

  setSelection(ids: Iterable<number>) {
    this.selected = new Set(ids);
  }

  /** Returns true when the object was ADDED (so the host can pop it). */
  toggle(oid: number): boolean {
    if (this.selected.has(oid)) {
      this.selected.delete(oid);
      return false;
    }
    this.selected.add(oid);
    return true;
  }

  hasSeg(): boolean {
    return this.masks.size > 0 || this.pointMask !== null;
  }

  active(): boolean {
    return !!(this.hoverId || this.selected.size || this.pointMask !== null);
  }

  buildMorph(oldId: number, newId: number): Morph | null {
    const fr = this.rings.get(oldId) ?? null;
    const to = this.rings.get(newId) ?? null;
    if (fr === null && to === null) return null;
    return new Morph(
      fr,
      to,
      this.colors.get(oldId) ?? this.colors.get(newId) ?? POINT_COLOR,
      this.colors.get(newId) ?? this.colors.get(oldId) ?? POINT_COLOR,
      this.secPolys.get(oldId) ?? null,
      this.secPolys.get(newId) ?? null,
      this.centroids.get(oldId) ?? null,
      this.centroids.get(newId) ?? null,
    );
  }

  scene(image: Handle | null, imageSize: Vec2): Scene {
    const sc = new Scene();
    sc.image = image;
    sc.imageSize = imageSize;
    sc.segMode = this.segMode;
    sc.masks = this.masks;
    sc.colors = this.colors;
    sc.polys = this.polys;
    sc.secPolys = this.secPolys;
    sc.centroids = this.centroids;
    sc.radius = this.radius;
    sc.selected = this.selected;
    sc.hoverId = this.hoverId;
    sc.hoverGen = this.hoverGen;
    sc.hoverSpec = this.hoverSpec;
    sc.hoverDepth = this.hoverDepth;
    sc.pointMask = this.pointMask;
    sc.pointPolys = this.pointPolys;
    sc.morph = this.morph;
    sc.clipActive = this.clipActive;
    sc.clipMasks = this.clipMasks;
    sc.clipBg = this.clipBg;
    return sc;
  }
}
